pyop2/sequential.py

```diff
@@ -1096,20 +1096,20 @@ def extrusion_loop():
             continue
         _buf_name[arg] = "buffer_%s" % arg.c_arg_name(count)
         _tmp_name[arg] = "tmp_%s" % _buf_name[arg]
-        _buf_size = list(itspace._extents)
+        _loop_size = [m.arity for m in arg.map]
         if not arg._is_mat:
             # Readjust size to take into account the size of a vector space
-            _dat_size = (arg.data.cdim,)
-            _buf_size = [sum([e*d for e, d in zip(_buf_size, _dat_size)])]
-            _loop_size = [_buf_size[i]//_dat_size[i] for i in range(len(_buf_size))]
+            _dat_size = [_arg.data.cdim for _arg in arg]
+            _buf_size = [sum([e*d for e, d in zip(_loop_size, _dat_size)])]
         else:
             _dat_size = arg.data.dims[0][0]  # TODO: [0][0] ?
-            _buf_size = [e*d for e, d in zip(_buf_size, _dat_size)]
+            _buf_size = [e*d for e, d in zip(_loop_size, _dat_size)]
```

Comment:
Can you explain what is what here, and why does `_buf_size` change to `_loop_size`?

Reply:
The buffer sizes seem wrong form many cases, e.g. `tests/regression/test_2dcohomology.py::test_betti1[space5]`

The generated wrapper is
```C
double buffer_arg0_0[20] __attribute__((aligned(16))) = {0.0};
double buffer_arg1_0[40] __attribute__((aligned(16)));
double buffer_arg2_0[20] __attribute__((aligned(16)));
double buffer_arg3_0[40] __attribute__((aligned(16)));
```
whereas the correct sizes are
```C
double buffer_arg0_0[20] __attribute__((aligned(16))) = {0.0};
double buffer_arg1_0[6] __attribute__((aligned(16)));
double buffer_arg2_0[20] __attribute__((aligned(16)));
double buffer_arg3_0[30] __attribute__((aligned(16)));
```
As far as I see, it seems to be a consequence of computing the buffer sizes from iteration extents.